examples/with-apollo/lib/with-apollo.js

```diff
@@ -41,31 +66,59 @@ const withApollo = PageComponent => {
         // getDataFromTree does not call componentWillUnmount
         // head side effect therefore need to be cleared manually
         Head.rewind()
-      }
 
-      // Extract query data from the Apollo store
-      const apolloState = apolloClient.cache.extract()
+        // Extract query data from the Apollo store
+        const apolloState = apolloClient.cache.extract()
 
-      return {
-        ...pageProps,
-        apolloState
+        return {
+          ...pageProps,
+          apolloState
+        }
       }
     }
+  }
 
-    constructor (props) {
-      super(props)
-      this.apolloClient = props.apolloClient || initApollo(props.apolloState)
-    }
+  return WithApollo
+}
 
-    render () {
-      const { apolloClient, apolloState, ...pageProps } = this.props
-      return (
-        <ApolloProvider client={this.apolloClient}>
-          <PageComponent {...pageProps} />
-        </ApolloProvider>
-      )
-    }
+/**
+ * Always creates a new apollo client on the server
+ * Creates or reuses apollo client in the browser.
+ * @param  {Object} initialState
+ */
+function initApollo (initialState) {
+  // Make sure to create a new client for every server-side request so that data
+  // isn't shared between connections (which would be bad)
+  if (typeof window === 'undefined') {
+    return createApolloClient(initialState)
   }
+
+  // Reuse client on the client-side
+  if (!apolloClient) {
+    apolloClient = createApolloClient(initialState)
+  }
+
+  return apolloClient
+}
+
+/**
+ * Creates and configures the ApolloClient
+ * @param  {Object} [initialState={}]
+ */
+function createApolloClient (initialState = {}) {
+  // Check out https://github.com/zeit/next.js/pull/4611 if you want to use the AWSAppSyncClient
+  const isBrowser = typeof window !== 'undefined'
+  return new ApolloClient({
+    connectToDevTools: isBrowser,
+    ssrMode: !isBrowser, // Disables forceFetch on the server (so queries are only run once)
+    link: new HttpLink({
+      uri: 'https://api.graph.cool/simple/v1/cixmkt2ul01q00122mksg82pn', // Server URL (must be absolute)
+      credentials: 'same-origin', // Additional fetch() options like `credentials` or `headers`
+      // Use fetch() polyfill on the server
+      fetch: !isBrowser && fetch
+    }),
+    cache: new InMemoryCache().restore(initialState)
+  })
 }
 
 export default withApollo
```

Comment:
@HaNdTriX thank you for this PR! WDYT of naming the remaining file to `lib/apollo.js` and exposing `withApollo` as a normal (no-default export) export. Import would look more conventional this way. It will also be consistent to export something like `getApolloClient()` function from this file, which might be needed in some edge cases.

Higher order component is a rare pattern these days because of react hooks, so I guess it's fine to not follow an old convention, where a HOC is expected to be a default export in a file with a matching name.

Reply:
I already thought about that, too. Thanks for your feedback! 🙏✅